List database and print fail trace

dfuchsgruber opened the pull request:
<!-- 
Thank you for contributing a pull request!
Please name and describe your PR as you would write a
commit message.
-->


### What does this implement/fix?
This PR adds two convenience functions to SEML:

#### Listing
`seml list [pattern]` will list all collections in the database that matches an (optional) RegEx pattern. It will also print the state counts for all experiments in the collection. Note that in contrast to `seml {collection} status`, it does not verify that experiments with state `RUNNING` are actually still running and have not died or been interupted. This is to ensure the command runs quickly.

Example output:
```bash
seml list .*_pneuma
WARNING: Status of RUNNING experiments may not reflect if they have died or been canceled. Use `seml ... status` instead.
+------------+--------+---------+---------+--------+--------+-------------+-----------+-------+
| Collection | STAGED | PENDING | RUNNING | FAILED | KILLED | INTERRUPTED | COMPLETED | Total |
+------------+--------+---------+---------+--------+--------+-------------+-----------+-------+
| gat_pneuma |      0 |       0 |       0 |      7 |      0 |           0 |        65 |    72 |
| gcn_pneuma |      0 |       0 |       2 |      1 |      0 |           0 |        45 |    48 |
| hpo_pneuma |      0 |       0 |       0 |     24 |     49 |           0 |        47 |   120 |
+------------+--------+---------+---------+--------+--------+-------------+-----------+-------+
```
#### Fail Trace Printing
`seml {collection} print-fail-trace` will print the fail trace of all experiments with "failed states" (`FAILED`, `KILLED`, `INTERRUPTED`). This way, you don't have to use a MongoDB client everytime you want to inspect error logs.

Example output (shortened, to only show the output format):
```bash
seml gat_pneuma print-fail-trace
***** Experiment ID 1, status: FAILED, slurm array-id, task-id: 7921342-0 *****
        Traceback (most recent call last):
          File "/nfs/staff-ssd/fuchsgru/miniconda3/envs/trajectory_gnn/lib/python3.10/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 92, in _call_target
    ...
TypeError("new() received an invalid combination of arguments - got (str, int), but expected one of:\n * (*, torch.device device)\n      didn't match because some of the arguments have invalid types: (\x1b[31;1mstr\x1b[0m, \x1b[31;1mint\x1b[0m)\n * (torch.Storage storage)\n * (Tensor other)\n * (tuple of ints size, *, torch.device device)\n * (object data, *, torch.device device)\n")
full_key: model.layers0

***** Experiment ID 35, status: FAILED, slurm array-id, task-id: 7930611-2 *****
        Traceback (most recent call last):
          File "/nfs/homedirs/fuchsgru/seml/seml/hydra.py", line 98, in decorator
    result = func(cfg)
          ...
        torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 1.75 GiB (GPU 0; 39.59 GiB total capacity; 37.05 GiB already allocated; 1.21 GiB free; 37.14 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF

...
```

n-gao commented:
Thanks a lot for the great contribution! I did not yet have the time to fully review but here are a few things I noticed.
For `seml list`:
* `seml list` fails if a collection is not created by seml or contains non-seml items.
* In consistency with the rest of the code base and for better configurability, we use `logging.info` instead of `print`
* Should we use [aggregate sum of PyMongo](https://www.mongodb.com/docs/manual/reference/operator/aggregation/sum/) instead of summing in python? That would reduce the network load and offload more to the database which typically is the most efficient way.
* Should we just use `pandas` for pretty printing since we already have it as a requirement?
* For backward and future compatibility, we should iterative over all aliases for the same state rather than taking only the first one.

For `print-fail-trace`:
* `logging.info` instead of `print`
* Probably add a final message saying how many experiments have been crawled and how many have been printed. This is to catch typos in the collection name since atm the program just exits without any message.

dfuchsgruber commented:
Thanks for the feedback, I hope I addressed your points. Remaining remarks:

> * Should we just use `pandas` for pretty printing since we already have it as a requirement?

I implemented it this way now and removed the prettytable dependency, but imho I find the result visually less appealing. A matter of taste in the end. It now looks like this:
```bash
seml list .*pneuma
WARNING: Status of RUNNING experiments may not reflect if they have died or been canceled. Use `seml ... status` instead.
                STAGED  PENDING  RUNNING  FAILED  KILLED  INTERRUPTED  COMPLETED  Total
Collection                                                                             
gat_pneuma           0        0        0       7       0            0         65     72
gcn_pneuma           0        0        2       1       0            0         45     48
hpo_pneuma           0        0        0      24      49            0         47    120
pneuma               0        0        1       0       0            0          9     10
pneuma_spatial       0        0        0       2       1            0         13     16
```

n-gao commented on seml/database.py:
```diff
@@ -3,11 +3,15 @@
 import logging
 from tqdm.auto import tqdm
+import re
+from collections import Counter
```
We technically do not need the Counter anymore, but you use it here as a `defaultdict`.

n-gao commented on seml/database.py:
```diff
@@ -277,3 +281,40 @@ def clean_unreferenced_artifacts(db_collection_name=None, yes=False):
+    db = get_database(**mongodb_config)
+    collection_names = [name for name in db.list_collection_names()
+                        if name not in ('fs.chunks', 'fs.files') and re.compile(pattern).match(name)]
+    name_to_counts = {}
```
To remove the counter usage:
`name_to_counts = defaultdict(lambda: defaultdict(lambda: 0))`
and later:
`name_to_counts[collection_name].update({result['_id'] : result['_count'] for result in counts_by_status})`

dfuchsgruber replied:
What's the issue with Counters? Isn't `Counter` specifically developed as a nice way to do `defaultdict(lambda: 0)`?

n-gao replied:
That's an interesting take on counter. Never thought about it that way ^^ while functionally true but wording wise it seems a bit confusing as I at least first thought that something is counted there

n-gao replied:
But we see each status only exactly once since the counting has been done by mongodb. So it's technically just a defaultdict?

dfuchsgruber replied:
True, but still the (default)dict represents what you'd normally represent with a counter. I always use this data structure for dicts that are like that, mostly because of convenience functions (which we do not need here), consistency and readability. But I also don't wanna get lost in nitpicky stuff, we can also use defaultdicts.

n-gao commented on seml/database.py:
```diff
@@ -277,3 +281,40 @@ def clean_unreferenced_artifacts(db_collection_name=None, yes=False):
+        counts_by_status = db[collection_name].aggregate([{'$group' : {'_id' : '$status', '_count' : {'$sum' : 1}}}])
+        name_to_counts[collection_name] = Counter({result['_id'] : result['_count'] for result in counts_by_status})
+    
+    columns = [States.STAGED, States.PENDING, States.RUNNING, States.FAILED, States.KILLED, States.INTERRUPTED, 
```
wouldn't this just be `States.keys()`?

n-gao replied:
Since python 3.6 insertion order of dicts is preserved and this is touted as a feature not a mere implementation. Thus, we can just take the same order as in settings.py with less redundancy on the ordering.

n-gao commented on seml/database.py:
```diff
@@ -277,3 +281,40 @@ def clean_unreferenced_artifacts(db_collection_name=None, yes=False):
+               States.COMPLETED]
+    table = []
+    for name, counts in name_to_counts.items():
+        if list_empty or any(sum(counts[state] for state in states) > 0 for states in columns):
```
Okay, this comment is now roughly 30min later, and a bunch of unnecessary testing later. I did not super like the long if statement, the reiteration of the list, and the misuse of a counter. So, I thought: can't we just directly use a dataframe. And the answer is yes:
```python
    logging.warning(f"Status of {States.RUNNING[0]} experiments may not reflect if they have died or been canceled. Use `seml ... status` instead.")
    if mongodb_config is None:
        mongodb_config = get_mongodb_config()
    db = get_database(**mongodb_config)
    collection_names = [name for name in db.list_collection_names()
                        if name not in ('fs.chunks', 'fs.files') and re.compile(pattern).match(name)]
    name_to_counts = defaultdict(lambda: {state: 0 for state in States.keys()})
    it = tqdm(collection_names) if progress else collection_names
    
    inv_states = {v: k for k, states in States.items() for v in states}
    for collection_name in it:
        counts_by_status = db[collection_name].aggregate([{'$group' : {'_id' : '$status', '_count' : {'$sum' : 1}}}])
        name_to_counts[collection_name].update({
            inv_states[result['_id']]: result['_count']
            for result in counts_by_status
            if result['_id'] in inv_states
        })
    
    df = pd.DataFrame.from_dict(name_to_counts, dtype=int).transpose()
    # Remove empty collections
    if not list_empty:
        df = df[df.sum(axis=1) > 0]
    # add a column with the total
    df['Total'] = df.sum(axis=1)
    logging.info(df.sort_index().to_string())
```
This is purely to please my own ego and to shorten a long line of codes and I definitely don't want to diminish your work!

dfuchsgruber replied:
Personally, I think this is less readable, which, for non-runtime critical applications like seml, I prefer to be the maxime. However, if you want this version, I have no objections, I only care about the feature being realized.

n-gao replied:
Interesting, I thought of it being more readable ^^ but that may be because I authored it or because of the comments. But this implementation also just takes the ordering from settings.py which may be desired(?).

dfuchsgruber replied:
As I said, take yours if you prefer it, the ordering from settings.py is definitely a good point that one can and should realize.

n-gao reviewed (commented):
Thanks for addressing my feedback! I left some comments for the listing of databases, for the failtrace printing, why does it print for interrupted or killed experiments? Shouldn't they have no stack trace?
